storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,22 +3571,401 @@ dberr_t fsp_sys_tablespace_validate()
+      }
+    }
+  }
+
+  if (err == DB_SUCCESS)
+    err= fsp_table_inodes(inodes, mtr);
+  return err;
+}
+
+/** Free the extents, fragment page from the given inode
+@param page_no index node page number
+@param offset  index node offset within page
+@return error code */
+static void fseg_inode_free(uint32_t page_no, uint16_t offset)
+{
+  fil_space_t *space= fil_system.sys_space;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  sql_print_information("InnoDB: Freeing the unused segment page "
+                        UINT32PF " offset " UINT32PF, page_no,
+                        offset);
+  buf_block_t *iblock= buf_page_get_gen(page_id_t{0, page_no}, 0,
+                                        RW_X_LATCH, nullptr, BUF_GET,
+                                        &mtr);
+
+  DBUG_EXECUTE_IF("unused_undo_free_fail_4", iblock= nullptr;);
+  if (!iblock)
+  {
+    mtr.commit();
+    return;
+  }
+
+  fseg_inode_t *inode= iblock->page.frame + offset;
+  bool all_pages_freed= false;
+  while (!fseg_free_step_low(iblock, inode, space, &mtr, nullptr,
+                             &all_pages_freed))
+  {
+    DBUG_EXECUTE_IF("unused_undo_free_fail_5",
+                    mtr.commit();
+                    return;);
+    iblock->fix();
+    mtr.commit();
+
+    mtr.start();
+    mtr.x_lock_space(space);
+    iblock->page.lock.x_lock();
+    mtr.memo_push(iblock, MTR_MEMO_PAGE_X_FIX);
+  }
+
+  DBUG_EXECUTE_IF("unused_undo_free_fail_6", mtr.commit(); return;);
+  if (all_pages_freed)
+    /* Freeing completed: free the segment inode */
+    fsp_free_seg_inode(space, inode, iblock, &mtr);
+  mtr.commit();
+  return;
+}
+
+/** Free the unused segment
+@param boffset    offset of the list containing segment header
+@param space      system tablespace
+@param usedInodes Used system tablespace index node */
+static void fsp_free_unused_seg(inode_info *unused_inodes)
+{
+  for (auto it= unused_inodes->start();
+       it != unused_inodes->last(); it++)
+    fseg_inode_free((uint32_t)((*it) >> 32), (uint16_t)(*it));
+}
+
+static dberr_t fsp_get_sys_unused_segment(inode_info *inodes,
+                                          uint16_t boffset,
+                                          inode_info *unused_inodes)
+{
+  dberr_t err= DB_SUCCESS;
+  buf_block_t *block= buf_pool.page_fix(page_id_t{0, 0}, &err,
+                                        buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+  buf_block_t *header= block;
+  const uint32_t len= flst_get_len(block->page.frame + boffset);
+  fil_addr_t addr= flst_get_first(block->page.frame + boffset);
+  for (uint32_t i= len; i-- && err == DB_SUCCESS; )
+  {
+    ut_ad(addr.boffset >= FIL_PAGE_DATA);
+    ut_ad(addr.boffset < block->physical_size() - FIL_PAGE_DATA_END);
+    block= buf_pool.page_fix(page_id_t{0, addr.page}, &err,
+                             buf_pool_t::FIX_WAIT_READ);
+    if (!block)
+    {
+      addr.page= FIL_NULL;
+      continue;
+    }
```

Comment:
It would be nice to add a comment that explains why we would tolerate this error.

Reply:
It was added to avoid the assert. This has been fixed now.